exploration/variable-mutability.md

```diff
@@ -0,0 +1,244 @@
+  > let $foo = {$foo :trim}
+  > let $foo = {$foo :sanitize target=html}
+  > ```
+  >
+  > This can also be achieved by renaming:
+  >
+  > ```
+  > let $foo1 = {$foo :text-transform transform=uppercase}
+  > let $foo2 = {$foo1 :trim}
+  > let $foo3 = {$foo2 :sanitize target=html}
+  > ```
+- Users want to impose typing on (we say "annotate") external variables
+  or literals:
+  > ```
+  > let $fooAsNumber = {$foo :number}
+  > let $anotherNumber = {42 :number}
+  > ```
+- Users may wish to provide complex annotations which are reused across mulitple patterns
+
+  > ```
+  > let $count = {$count :number}
+  > let $date = {$date :datetime dateStyle=long}
+  > match {$count}
+  > when 1 {You received one message on {$date}}
+  > when * {You received {$count} messages on {$date}}
+  > ```
+
+- Implementers need to know what value is associated with a named variable, see #299.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- Be able to re-annotate variables without having to rename them in the message body
+- Allow static analysis to detect mistakes when referencing an undefined local variable
+- Be able to re-annotate variables multiple times (because we do not allow nesting)
+
+- _more needed?_
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+- Variable names are potentially contrained by `Nmtoken`.
+  The reason we chose Nmtoken (and Name) was to maximize compatibility with (potential)
+  LDML constructs, since CLDR uses XML.
+  The "carve out" for various sigils doesn't conflict with naming in LDML currently and
+  future conflicts are under our (CLDR-TC's) control.
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Separate local variables from externally passed values by altering the sigil
+and by using a visually distinctive pattern for local names
+(in an effort to prevent `$foo`/`@foo` confusion).
+
+```abnf
+variable     = local_var / external_var
+local_var    = "@_" name
+external_var = "$" name
+```
+
+> _Example_
+>
+> ```
+> let @_local = {$external :transform}
+> let @_anotherLocalVar = {|Some literal| :annotated}
+> ```
+
+To allow users to perform multiple annotations on a value,
+while still allowing detection of unintentional reassignment,
+introduce a new keyword `modify`:
+
+```abnf
+declaration = (let / modify) s variable [s] "=" [s] expression
+...
+modify = %6D.%6F.%64.%69.%66.%79
+```
+
+It is a syntax error to use `let` on a variable that has been previously
+assigned through any declaration (either `let` or `modify`)
+
+It is a variable resolution error to call `modify`
+on an external variable that does not exist.
+
+> _Example_
+>
+> ```
+> let @_local = {$external :transform}
+> modify @_local = {@_local :modification with=options}
+> modify $external = {$external :transform adding=annotation}
```

Comment:
This looks an awful lot like we're doing pass-by-reference for formatting function arguments and we're modifying the `$external` value in the calling context. Is that intentional?

Reply:
It intentionally **_looks like_** it is modifying the value. In practice, it (probably) does not have write access to the external value and is only masking it for the duration of this message. But the idea is that the value is mutable only if you are consciously mutating it.

Note that we could disallow `modify` to externals, that is, you have to either `let` or `modify` a local variable in order to annotate an external var.